source/extensions/common/wasm/plugin.cc

```diff
@@ -20,6 +16,44 @@ WasmConfig::WasmConfig(const envoy::extensions::wasm::v3::PluginConfig& config)
     // TODO(rapilado): Set the SanitizationConfig fields once sanitization is implemented.
     allowed_capabilities_[capability.first] = proxy_wasm::SanitizationConfig();
   }
+
+  if (config_.vm_config().has_environment_variables()) {
+    auto envs = config_.vm_config().environment_variables();
+
+    // We reject NullVm with environment_variables configuration
+    // since it directly accesses Envoy's env vars.
```

Comment:
@PiotrSikora is this fully necessary? Out of curiosity, how many places do we have different behaviors or checks for NullVm vs. sandboxed?

The reason I ask is that we've been thinking it is pretty easy to take Wasm and turn it into a C++ extension as NullVm (as you know in some discussions we've had in other forums). That is less likely to be true if we have many special cases like this.

Reply:
I believe it is, unfortunately, see my previous comment: https://github.com/envoyproxy/envoy/pull/15136#discussion_r586758400.

Looking at it now, I think we could allow use of `host_env_keys` (since it's a read-only access) and reject only configurations with custom `key_values`. @mathetake do you want to make that change?

Regarding "real" Proxy-Wasm C++ plugins compiled to NullVM (this only works one-way, btw), you're right, this would be the first thing that's available only inside the WasmVM sandbox... But we could expose the same key-value pairs via `proxy_get_map_values(type::EnvironmentVariables, ..)`, and then make sure that portable Proxy-Wasm C++ plugins don't use `std::env::var()` and call `proxy_get_map_values()` instead.

Reply:
>  But we could expose the same key-value pairs via proxy_get_map_values(type::EnvironmentVariables, ..), and then make sure that portable Proxy-Wasm C++ plugins don't use std::getenv() and call proxy_get_map_values() instead.

Good Idea, but if that is the case, we could expose both host_env_keys and key_values to NullVm through that call without modifying Envoy's envs, so we do not need to reject NullVm at all, right? 

That said, this time how about having comment like `TODO: once proxy_get_map_values(type::EnvironmentVariables, ..) call is supported, then remove the rejection of NullVm`? WDYT?

Reply:
> Good Idea, but if that is the case, we could expose both host_env_keys and key_values to NullVm through that call without modifying Envoy's envs, so we do not need to reject NullVm at all, right?

Correct. But maybe let's relax this restriction and only forbid `key_values` for now, until the generic `proxy_get_map_values` is implemented.

> That said, this time how about having comment like `TODO: once proxy_get_map_values(type::EnvironmentVariables, ..) call is supported, then remove the rejection of NullVm`? WDYT?

Yes, I'm fine with the `TODO`, since we don't have generic `proxy_get_map_values` in the current iteration of the ABI.

Reply:
OK, TODO is fine.